janus_core/calculations/eos.py

```diff
@@ -159,111 +53,244 @@ def calc_eos(
-    if n_volumes <= 1:
-        raise ValueError("`n_volumes` must be greater than 1.")
-    if not 0 < min_volume < 1:
-        raise ValueError("`min_volume` must be between 0 and 1.")
-    if max_volume <= 1:
-        raise ValueError("`max_volume` must be greater than 1.")
-
-    if minimize:
-        if logger:
-            logger.info("Minimising initial structure")
-            minimize_kwargs["log_kwargs"] = {
-                "filename": log_kwargs["filename"],
-                "name": logger.name,
-                "filemode": "a",
-            }
-        optimize(struct, **minimize_kwargs)
-
-        # Optionally write structure to file
-        output_structs(
-            images=struct, write_results=write_structures, write_kwargs=write_kwargs
+        self.write_kwargs.setdefault(
+            "filename",
+            self._build_filename("generated.extxyz").absolute(),
         )
 
-    # Set constant volume for geometry optimization of generated structures
-    if "filter_kwargs" in minimize_kwargs:
-        minimize_kwargs["filter_kwargs"]["constant_volume"] = True
-    else:
-        minimize_kwargs["filter_kwargs"] = {"constant_volume": True}
-
-    lattice_scalars, volumes, energies = _calc_volumes_energies(
-        struct=struct,
-        min_volume=min_volume,
-        max_volume=max_volume,
-        n_volumes=n_volumes,
-        minimize_all=minimize_all,
-        minimize_kwargs=minimize_kwargs,
-        write_structures=write_structures,
-        write_kwargs=write_kwargs,
-        logger=logger,
-        tracker=tracker,
-    )
-
-    if write_results:
-        with open(f"{file_prefix}-eos-raw.dat", "w", encoding="utf8") as out:
-            print("#Lattice Scalar | Energy [eV] | Volume [Å^3] ", file=out)
-            for eos_data in zip(lattice_scalars, energies, volumes):
-                print(*eos_data, file=out)
-
-    eos = EquationOfState(volumes, energies, eos_type)
-
-    if logger:
-        logger.info("Starting of fitting equation of state")
-        tracker.start_task("Fit EoS")
-
-    v_0, e_0, bulk_modulus = eos.fit()
-    # transform bulk modulus unit in GPa
-    bulk_modulus *= 1.0e24 / kJ
-
-    if logger:
-        tracker.stop_task()
-        tracker.stop()
-        logger.info("Equation of state fitting complete")
-
-    if write_results:
-        with open(f"{file_prefix}-eos-fit.dat", "w", encoding="utf8") as out:
-            print("#Bulk modulus [GPa] | Energy [eV] | Volume [Å^3] ", file=out)
-            print(bulk_modulus, e_0, v_0, file=out)
-
-    results = {
-        "eos": eos,
-        "bulk_modulus": bulk_modulus,
-        "e_0": e_0,
-        "v_0": v_0,
-    }
-
-    return results
+        if (
+            (self.minimize or self.minimize_all)
+            and "write_results" in self.minimize_kwargs
+            and self.minimize_kwargs["write_results"]
+        ):
+            raise ValueError(
+                "Optimized structures can be saved by setting `write_structures`"
```

Comment:
The idea is `write_structures` is a settable attribute of `EoS`, which controls this, rather than passing it through `minimize_kwargs["write_results"]`.

We could say `EoS.write_structures`, although ideally we could use this to catch errors via the CLI too, 
 which would be `--write-structures`.

Perhaps 

```python
raise ValueError(
    "Please set the `write_structures` parameter to `True` to save"
    "optimized structures, instead of passing `write_results` through "
    "`minimize_kwargs`"
```
?

Reply:
That would at least explain to me as an oblivious user what to do and why.